File: app/server/notify/channels/dingtalk.ts
/**
 * 钉钉机器人推送渠道
 */
import crypto from 'crypto';
import Logger from '../../utils/logger';
import { NotifyChannel, NotifyResult } from '../types';
import { $ } from '../httpClient';
import { hasConfig, getConfig } from '../config';

const logger = Logger.child({ channel: 'dingtalk' });

export const dingtalkChannel: NotifyChannel = {
    name: 'dingtalk',
    enabled: false,

    async send(text: string, desp: string): Promise<NotifyResult> {
        return new Promise((resolve) => {
            if (!hasConfig('DD_BOT_TOKEN')) {
                resolve({ success: false, message: 'DD_BOT_TOKEN 未配置' });
                return;
            }

            const DD_BOT_TOKEN = getConfig('DD_BOT_TOKEN');
            const DD_BOT_SECRET = getConfig('DD_BOT_SECRET');

            let url = `https://oapi.dingtalk.com/robot/send?access_token=${DD_BOT_TOKEN}`;

            // 签名
            if (DD_BOT_SECRET) {
                const now = Date.now();
                const stringToSign = `${now}\n${DD_BOT_SECRET}`;
                const sign = crypto
                    .createHmac('sha256', DD_BOT_SECRET)
                    .update(stringToSign)
                    .digest('base64');
                url += `&timestamp=${now}&sign=${encodeURIComponent(sign)}`;
            }

            const title = text.substring(0, 64);
            const content = desp.substring(0, 15000);

            const body = {
                msgtype: 'markdown',
                markdown: {
                    title,
                    text: content
                }
            };

            $.post({ url, json: body }, (err, _res, resBody) => {
                if (err) {
                    logger.error({ err }, '钉钉发送失败');
                    resolve({ success: false, error: err as Error });
                    return;
                }

                const data = resBody as { errcode?: number; errmsg?: string };
                if (data?.errcode === 0) {
                    logger.info('钉钉发送成功');
                    resolve({ success: true, message: '发送成功' });
                } else {
                    logger.warn({ body: resBody }, '钉钉发送失败');
                    resolve({ success: false, message: data?.errmsg || '发送失败' });
                }
            });
        });
    }
};

// 检查是否启用
if (hasConfig('DD_BOT_TOKEN')) {
    dingtalkChannel.enabled = true;
}

export default dingtalkChannel;

File: app/server/internal/notify/channels/dingtalk.go
package channels

import (
	"crypto/hmac"
	"crypto/sha256"
	"encoding/base64"
	"fmt"
	"log/slog"
	"net/url"
	"time"

	"github.com/sushazhi/fnos-logmanager/internal/notify"
)

type DingTalk struct {
	enabled bool
}

func (c *DingTalk) Name() string {
	return "dingtalk"
}

func (c *DingTalk) Enabled() bool {
	return c.enabled
}

func (c *DingTalk) SetEnabled(enabled bool) {
	c.enabled = enabled
}

func (c *DingTalk) Send(text, desp string) notify.NotifyResult {
	if !notify.HasConfig("DD_BOT_TOKEN") {
		return notify.NotifyResult{Success: false, Message: "DD_BOT_TOKEN 未配置"}
	}

	ddBotToken := notify.GetConfig("DD_BOT_TOKEN")
	ddBotSecret := notify.GetConfig("DD_BOT_SECRET")

	webhookURL := "https://oapi.dingtalk.com/robot/send?access_token=" + ddBotToken

	// HMAC-SHA256 signature
	if ddBotSecret != "" {
		now := time.Now().UnixMilli()
		stringToSign := fmt.Sprintf("%d\n%s", now, ddBotSecret)
		mac := hmac.New(sha256.New, []byte(ddBotSecret))
		mac.Write([]byte(stringToSign))
		sign := base64.StdEncoding.EncodeToString(mac.Sum(nil))
		webhookURL += fmt.Sprintf("&timestamp=%d&sign=%s", now, url.QueryEscape(sign))
	}

	// Truncate title to 64 chars and content to 15000 chars
	title := text
	if len(title) > 64 {
		title = title[:64]
	}
	content := desp
	if len(content) > 15000 {
		content = content[:15000]
	}

	body := map[string]interface{}{
		"msgtype": "markdown",
		"markdown": map[string]string{
			"title": title,
			"text":  content,
		},
	}

	resp, err := notify.HTTPPost(webhookURL, notify.HttpRequestOptions{
		JSON: body,
	})
	if err != nil {
		slog.Error("钉钉发送失败", "error", err)
		return notify.NotifyResult{Success: false, Message: err.Error(), Err: err}
	}

	if resp.StatusCode == 200 {
		slog.Info("钉钉发送成功")
		return notify.NotifyResult{Success: true, Message: "发送成功"}
	}

	slog.Warn("钉钉发送失败", "status", resp.StatusCode, "body", resp.Body)
	return notify.NotifyResult{Success: false, Message: "发送失败"}
}

func init() {
	ch := &DingTalk{}
	if notify.HasConfig("DD_BOT_TOKEN") {
		ch.SetEnabled(true)
	}
	notify.Registry.Register(ch)
}
